homeassistant/components/tuya/__init__.py

```diff
@@ -125,59 +97,44 @@ async def cleanup_device_registry(
 
 async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Unloading the Tuya platforms."""
-    unload = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
-    if unload:
-        hass_data: HomeAssistantTuyaData = hass.data[DOMAIN][entry.entry_id]
-        hass_data.device_manager.mq.stop()
-        hass_data.device_manager.remove_device_listener(hass_data.device_listener)
-
-        hass.data[DOMAIN].pop(entry.entry_id)
-        if not hass.data[DOMAIN]:
-            hass.data.pop(DOMAIN)
+    if unload_ok := await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
+        tuya: HomeAssistantTuyaData = hass.data[DOMAIN][entry.entry_id]
+        if tuya.manager.mq is not None:
+            tuya.manager.mq.stop()
+        tuya.manager.remove_device_listener(tuya.listener)
+        await hass.async_add_executor_job(tuya.manager.unload)
+        del hass.data[DOMAIN][entry.entry_id]
+    return unload_ok
 
-    return unload
 
-
-class DeviceListener(TuyaDeviceListener):
+class DeviceListener(SharingDeviceListener):
     """Device Update Listener."""
 
     def __init__(
         self,
         hass: HomeAssistant,
-        device_manager: TuyaDeviceManager,
-        device_ids: set[str],
+        manager: Manager,
     ) -> None:
         """Init DeviceListener."""
         self.hass = hass
-        self.device_manager = device_manager
-        self.device_ids = device_ids
+        self.manager = manager
 
-    def update_device(self, device: TuyaDevice) -> None:
+    def update_device(self, device: CustomerDevice) -> None:
         """Update device status."""
-        if device.id in self.device_ids:
-            LOGGER.debug(
-                "Received update for device %s: %s",
-                device.id,
-                self.device_manager.device_map[device.id].status,
-            )
-            dispatcher_send(self.hass, f"{TUYA_HA_SIGNAL_UPDATE_ENTITY}_{device.id}")
-
-    def add_device(self, device: TuyaDevice) -> None:
+        LOGGER.debug(
+            "Received update for device %s: %s",
+            device.id,
+            self.manager.device_map[device.id].status,
+        )
+        dispatcher_send(self.hass, f"{TUYA_HA_SIGNAL_UPDATE_ENTITY}_{device.id}")
+
+    def add_device(self, device: CustomerDevice) -> None:
         """Add device added listener."""
         # Ensure the device isn't present stale
         self.hass.add_job(self.async_remove_device, device.id)
 
-        self.device_ids.add(device.id)
```

Comment:
How are already added devices tracked now?

Reply:
It doesn't. 

The main reason this existed is to dispatch only devices added to Home Assistant. However, all devices are always added.